tubular/imputers.py

```diff
@@ -25,30 +30,31 @@ class attribute, indicates whether transformer has been converted to polars/pand
 
     """
 
-    polars_compatible = False
+    polars_compatible = True
 
     FITS = False
 
-    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+    @nw.narwhalify
+    def transform(self, X: FrameT) -> FrameT:
         """Impute missing values with median values calculated from fit method.
 
         Parameters
         ----------
-        X : pd.DataFrame
+        X : FrameT
             Data to impute.
 
         Returns
         -------
-        X : pd.DataFrame
+        X : FrameT
             Transformed input X with nulls imputed with the median value for the specified columns.
 
         """
         self.check_is_fitted(["impute_values_"])
 
-        X = super().transform(X)
+        X = nw.from_native(super().transform(X))
 
         for c in self.columns:
```

Comment:
B: this isn't using the narwhal's API particularly as you index the frame in the same way as pandas. As they say in [the polars docs](https://docs.pola.rs/user-guide/migration/pandas/#key-syntax-differences):

> If your Polars code looks like it could be pandas code, it might run, but it likely runs slower than it should.

Use the `nw.col(name)` syntax instead:

```python
import pandas as pd
import polars as pl
import narwhals as nw

df_pd = pd.DataFrame({"a": [2, 4, None, 3, 5], "b": [2.0, 4.0, float("nan"), 3.0, 5.0]})
df_pl = pl.DataFrame({"a": [2, 4, None, 3, 5], "b": [2.0, 4.0, float("nan"), 3.0, 5.0]})
impute_values = {'a' : 1, 'b' : 2}

>>> cols = ['a', 'b']

@nw.narwhalify
def func(df):
    return df.with_columns(nw.col(c).fill_null(impute_values[c]) for c in cols)

>>> func(df_pl)
shape: (5, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ f64 │
╞═════╪═════╡
│ 2   ┆ 2.0 │
│ 4   ┆ 4.0 │
│ 1   ┆ NaN │
│ 3   ┆ 3.0 │
│ 5   ┆ 5.0 │
└─────┴─────┘


>>> func(df_pd)
     a    b
0  2.0  2.0
1  4.0  4.0
2  1.0  2.0
3  3.0  3.0
4  5.0  5.0

```

Reply:
(apologies, this should be a few lines down)

Reply:
actioned here https://github.com/lvgig/tubular/pull/333/commits/b995e2785a70d7082eef1367dcd745dc5ea1700d

Reply:
Sorry to be a stickler on this one, but this:

```python
for c in self.columns:
            X = X.with_columns(nw.col(c).fill_null(self.impute_values_[c]).alias(c))
```

Isn't the same as:

```python

X = X.with_columns(nw.col(c).fill_null(self.impute_values_[c]) for c in self.columns)
```

a.) You don't need the `alias(c)` bit - you only need that if you want [to rename something](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.alias.html).
b.) repeated `with_columns` calls have an overhead as polars can't parallelise operations on an eager frame (on a lazy frame it matters less as the query plan isn't constructed until the end).

If you feel that the above line is too long, you can construct the exprs separately:

```python

fill_null_from_fitted_exprs = [
    nw.col(c).fill_null(self.impute_values_[c]) for c in self.columns
]

X = X.with_columns(fill_null_from_fitted_exprs)
```

Reply:
cool, actioned here https://github.com/lvgig/tubular/pull/333/commits/ac84dcc38a2d8b5bf5e3d85fafeb82392c94a6a3